### ISSUE TITLE
Questions about propagate function and message function on the GAT conv layer

### ISSUE THREAD

**Issue by chunyuma** [rusty1s/pytorch_geometric#1949]
## ❓ Questions & Help

Hello @rusty1s, I have a few questions about `propagate` function and `message` function used in [GAT](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/nn/conv/gat_conv.html) model:

1. When we call `propagate` in the forward method, can we add some customized parameters that we could use in `message` function? For example, if I want to add the edge type information of edge (i,j) when the `alpha` is calculated for edge (i,j), can I pass one parameter, say `edge_matrix` where each row vector represents the edge type of edge (i,j) to the `propagate` function? If I do it, how can I access `edge_matrix` in the `message` function?

2. What is the connection logic between `propagate` and `message`? Based on my understanding (please correct me if I'm wrong), the `message` function only processes one edge (i,j) independently based on `edge_index` used in `propagate` function. So the parameters `x_j`, `alpha_j` and `alpha_i` correspond to the specific row in `x=(x_l,x_r)` and `alpha=(alpha_l, alpha_r)` based on the `edge_index`. However, the code `alpha = F.dropout(alpha, p=self.dropout, training=self.training)` makes me confused. Here, `alpha` should be an attention value but not a vector corresponding to one edge (i,j), right? Why the `dropout` was used for a value? 

3. What does the `index` used in `message` represent/correspond to? I know that the `softmax` function uses this `index` to normalizes the `alpha` across the same target nodes. But I have hard time understanding the underlying magic?
 
Thank you!

**Comment by rusty1s** [comment]
Hi!

1. Yes, this is possible. Everything passed to `self.propagate` can be accessed in `message`:
```python
self.propagate(edge_index, x=x, edge_matrix=edge_matrix)

def message(self, x_j, edge_matrix):
    # TODO
```
2. The `message` function will process all edges in parallel. For example, `x_j` is a `[num_edges, num_features]` tensor holding the node features of source nodes. Hence, the dropout on `alpha` leads to a dropout of incoming messages.
3. The `index` has special meaning and refers to the index of target nodes. It is equivalent to `edge_index[1]`.

**Comment by chunyuma** [comment]
Thanks for your prompt reply @rusty1s!

 > The message function will process all edges in parallel.

Regarding this statement, I'm still a little bit confused. Let say I have a simple bipartite graph:
```python
edge_index[0] = [1,2,3,2] # source nodes' indexes
edge_index[1] = [1,1,2,2] # target nodes' indexes
# x_l.shape is (3, 1, 10) representing (Node ,Head, Channel)
# x_r.shape is (2, 1, 10) representing (Node ,Head, Channel)
# alpha_l.shape is (3, 1, 1)
# alpha_l.shape is (2, 1, 1)
# edge_matrix = (4, 1, 1) representing (Edge, Head, Channel)
```
When you said the message function process all edges in parallel, can I understand that the first dim of `x_j`, `x_i`,  `alpha_j`, `alpha_i` and `edge_matrix` within message function below are all 4 in my case above? So they correspond to the `edge_index` and it is no problem that we can calculate like `alpha_j + alpha_i + edge_matrix`.

Also, the shape of output alpha in `alpha = softmax(alpha, index, ptr, size_i)` is still (4, 1, 1) and `alpha[0]+alpha[1]==1`, right?

```python
def message(self, x_j: Tensor, alpha_j: Tensor, alpha_i: OptTensor, edge_matrix: Tensor, index: Tensor, ptr: OptTensor, size_i: Optional[int]) -> Tensor:
    alpha = alpha_j if alpha_i is None else alpha_j + alpha_i + edge_matrix
    alpha = F.leaky_relu(alpha, self.negative_slope)
    alpha = softmax(alpha, index, ptr, size_i)
    self._alpha = alpha
    alpha = F.dropout(alpha, p=self.dropout, training=self.training)
    return x_j * alpha.unsqueeze(-1)
```

Finally, can I also know the answers for the following simple questions?

1. What is the difference between `Tensor` and `OptTensor`? Does `alpha_i` have to be `OptTensor`?
2. What is the meaning of `ptr`?
3. Is `adj_t` in message_and_aggregate function just an adjacency matrix or a transposed adjacency matrix?

Thank you so much!

**Comment by rusty1s** [comment]
Yes, they all have a shape of 4 in the first dimension, and `alpha` will sum up to 1 across neighborhoods.

1. `OptTensor` is an `Optional[Tensor]` (so a `Tensor` which can be `None`). `alpha_i` is an `OptTensor` because we allow `x_r` to be optional. For example, in a bipartite graph where target nodes do not have features.
2. `ptr` is the `rowptr` in a sparse matrix CSR format. This only has an effect in case you are using the `SparseTensor` format. For example, in case `index` is `[0, 0, 0, 1, 1, 2]`, then `ptr` is `[0, 3, 5, 6]`.
3. This is the transposed adjacency matrix. This is needed to guarantee that messages flow from source to target nodes (and not vice versa). In case your graph is undirected, that does not have any effect.

**Comment by chunyuma** [comment]
I see, thanks @rusty1s
